typing/typemod.ml

```diff
@@ -2587,18 +2587,37 @@ let maybe_infer_modalities ~loc ~env ~md_mode ~mode =
     Mode.Modality.Value.id
   end
 
-let rec type_module ?(alias=false) sttn funct_body anchor env smod =
+type alias =
+  | No
+  | Yes_walk_locks
+  | Yes_hold_locks
+
+let is_alias = function
+  | No -> false
+  | Yes_walk_locks | Yes_hold_locks -> true
+
+let rec type_module_maybe_hold_locks ~alias sttn funct_body anchor env smod =
   Builtin_attributes.warning_scope smod.pmod_attributes
     (fun () -> type_module_aux ~alias sttn funct_body anchor env smod)
 
+and type_module ?(alias=false) sttn funct_body anchor env smod =
```

Comment:
Define multiple versions of this function, and get rid of `fatal_error`.

Reply:
I cannot define mutliple versions of the function, because that would create too much code duplication. I instead use GADT which maybe isn't so nice either (and I'm happy to revert)

Reply:
Yeah, it pains me to say it but I don't think the GADT is helping on balance.